packages/contracts/contracts/tutorials/Arbitrage.sol

```diff
@@ -0,0 +1,103 @@
+pragma solidity ^0.4.19;
+pragma experimental ABIEncoderV2;
+
+import "../Exchange.sol";
+import "../tokens/Token.sol";
+import "./EtherDelta.sol";
+import "../tokens/WETH9.sol";
+
+contract Arbitrage is Ownable {
+
+    Exchange exchange;
+    EtherDelta etherDelta;
+    address proxyAddress;
+
+    uint constant MAX_UINT = 2**256 - 1;
+
+    function Arbitrage(address _exchangeAddress, address _etherDeltaAddress, address _proxyAddress) {
+        exchange = Exchange(_exchangeAddress);
+        etherDelta = EtherDelta(_etherDeltaAddress);
+        proxyAddress = _proxyAddress;
+    }
+
+    function setAllowances(address tokenAddress) public onlyOwner {
+        Token token = Token(tokenAddress);
+        token.approve(address(etherDelta), MAX_UINT);
+        token.approve(proxyAddress, MAX_UINT);
+        token.approve(owner, MAX_UINT);
+    }
+
+    /*
+     * I 愛 the limitations on Solidity stack size!
+     *
+     * addresses
+     * 0..4 orderAddresses
+     * 5 tokenGet
+     * 6 tokenGive
+     * 7 user
+     *
+     * values
+     * 0..5 orderValues
+     * 6 fillTakerTokenAmount
+     * 7 amountGet
+     * 8 amountGive
+     * 9 expires
+     * 10 nonce
+     * 11 amount
+
+     * signature
+     * exchange then etherDelta
+     */
+    function makeAtomicTrade(
+        address[8] addresses, uint[12] values,
+        uint8[2] v, bytes32[2] r, bytes32[2] s
+    ) public onlyOwner {
+        makeExchangeTrade(addresses, values, v, r, s);
+        makeEtherDeltaTrade(addresses, values, v, r, s);
+    }
+
+    function makeEtherDeltaTrade(
+        address[8] addresses, uint[12] values,
+        uint8[2] v, bytes32[2] r, bytes32[2] s
+    ) internal {
+        uint amount = values[11];
+        etherDelta.depositToken(addresses[5], values[7]);
```

Comment:
For the sake of a tutorial, I think automatically depositing before the EtherDelta trade and withdrawing after a trade is fine, but it should probably be optional if ever used in production (going to save a lot of gas by keeping tokens in ED).

Reply:
This code doesn't pretend to be gas efficient/safe/nice. It's just a tutorial.